### PR TITLE
fix(obstacle_avoidance_planner): adding missing functionality for stop margin due to out of drivable area

### LINKED ISSUE #4190: Adding missing PRs to obstacle_avoidance_planner since code refactoring
### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I've agreed with the maintainers that I can plan this task.

### Description

This task is to address the missing PRs in `obstacle_avoidance_planner` since code refactoring.

Refactoring PR : 
- https://github.com/autowarefoundation/autoware.universe/pull/2796#:~:text=docs%C2%A0%232963-,import%20missing%20PRs,-ee98754

Missing PRs : 
- https://github.com/autowarefoundation/autoware.universe/commit/ee9875496ff2636be8bf4a9ef1a861e984519489
- https://github.com/autowarefoundation/autoware.universe/commit/d6afa1bb7e8401586c97d9850f1caa06ac094ebd
- https://github.com/autowarefoundation/autoware.universe/commit/2d98ee84c9b422529a543922732e6f12046c0e20

### Purpose

Add the mentioned PRs to `obstacle_avoidance_planner`.

### Possible approaches

Integrate the missing pieces from the mentioned PRs into main branch

### Definition of done

- [ ] Create PR for each missing PR mentioned above
- [ ] Review & Merge PRs

### PR DESCRIPTION
**Pull request by ahmeddesokyebrahim**
## Description
This PR is to add missing functionality of stop margin due to out of drivable area in `obstacle_avoidance_planner` since refactoring done in #2796 

## Related links
closes #4190 
:exclamation: Must be merged with https://github.com/autowarefoundation/autoware_launch/pull/438 :exclamation: 

## Tests performed
Same as mentioned in original PR : #2786 

## Notes for reviewers
Original PR : #2786 
Refactoring PR : #2796 

## Interface changes
N.A.

## Effects on system behavior
Adding a margin for stop point when the the stop point is calculated for the footprint out of drivable area.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/obstacle_avoidance_planner/src/node.cpp [hunk 1]
```diff
@@ -438,15 +445,29 @@ void ObstacleAvoidancePlanner::insertZeroVelocityOutsideDrivableArea(
...
-        publishVirtualWall(traj_point.pose);
         return i;
       }
     }
     return std::nullopt;
   }();
 
   if (first_outside_idx) {
-    for (size_t i = *first_outside_idx; i < optimized_traj_points.size(); ++i) {
+    size_t stop_idx = *first_outside_idx;
```
Comment (takayuki5168) [inline planning/obstacle_avoidance_planner/src/node.cpp:455]:
```suggestion
    const size_t stop_idx = *first_outside_idx;
```

File: planning/obstacle_avoidance_planner/src/node.cpp [hunk 2]
```diff
@@ -438,15 +445,29 @@ void ObstacleAvoidancePlanner::insertZeroVelocityOutsideDrivableArea(
...
+    const auto & op_target_point = motion_utils::calcLongitudinalOffsetPoint(
```
Comment (takayuki5168) [inline planning/obstacle_avoidance_planner/src/node.cpp:456]:
```suggestion
    const auto op_target_point = motion_utils::calcLongitudinalOffsetPoint(
```

File: planning/obstacle_avoidance_planner/src/node.cpp [hunk 3]
```diff
@@ -438,15 +445,29 @@ void ObstacleAvoidancePlanner::insertZeroVelocityOutsideDrivableArea(
...
+    if (op_target_point) {
+      const auto target_point = op_target_point.get();
+      // confirm that target point doesn't overlap with the stop point outside drivable area
+      const auto dist =
+        tier4_autoware_utils::calcDistance2d(optimized_traj_points.at(stop_idx), target_point);
+      const double overlap_threshold = 1e-3;
+      if (dist > overlap_threshold) {
+        stop_idx = motion_utils::findNearestSegmentIndex(optimized_traj_points, target_point);
+      }
+    }
```
Comment (takayuki5168) [inline planning/obstacle_avoidance_planner/src/node.cpp:468]: Could you check if you can replace this code with the following function.
https://github.com/autowarefoundation/autoware.universe/blob/6bf539b88a33802676be7871d0492b81b25947bf/common/motion_utils/include/motion_utils/trajectory/trajectory.hpp#L1196-L1199

Comment (takayuki5168) [reply planning/obstacle_avoidance_planner/src/node.cpp:468]: @ahmeddesokyebrahim 
`insertStopPoint` calls `insertTargetPoint`, and the function also checks if a close point to the target point already exists in the trajectory.
By using this function, the code change on this PR will be much smaller which I prefer.
Let me know if you do not agree with mine.

Btw, can I merge the following PR first?
If merging this PR first will have trouble for you, I'll wait.
https://github.com/autowarefoundation/autoware.universe/pull/4112

Comment (ahmeddesokyebrahim) [reply planning/obstacle_avoidance_planner/src/node.cpp:468]: got your point, thanks for the clarification.
Then I will try it with your proposal.
And go ahead for sure with merging your PR firstly.

Comment (ahmeddesokyebrahim) [reply planning/obstacle_avoidance_planner/src/node.cpp]: @takayuki5168 : Please have a look with current implementation using `insertStopPoint()`.

File: planning/obstacle_avoidance_planner/src/node.cpp [hunk 4]
```diff
@@ -443,7 +449,17 @@ void ObstacleAvoidancePlanner::insertZeroVelocityOutsideDrivableArea(
...
+      const auto dist = tier4_autoware_utils::calcDistance2d(
+        optimized_traj_points.at(0), optimized_traj_points.at(stop_idx));
+      const auto dist_with_margin = dist - vehicle_stop_margin_outside_drivable_area_;
+      const auto first_outside_idx_with_margin =
+        motion_utils::insertStopPoint(dist_with_margin, optimized_traj_points);
+      if (first_outside_idx_with_margin) {
+        stop_idx = *first_outside_idx_with_margin;
+      }
+      publishVirtualWall(optimized_traj_points.at(stop_idx).pose);
+      for (size_t i = stop_idx; i < optimized_traj_points.size(); ++i) {
```
Comment (takayuki5168) [inline planning/obstacle_avoidance_planner/src/node.cpp:462]: Could you make variables private like this.
```cpp
      const auto stop_idx = [&]() {
        const auto dist = tier4_autoware_utils::calcDistance2d(
          optimized_traj_points.at(0), optimized_traj_points.at(*first_outside_idx));
        const auto dist_with_margin = dist - vehicle_stop_margin_outside_drivable_area_;
        const auto first_outside_idx_with_margin =
          motion_utils::insertStopPoint(dist_with_margin, optimized_traj_points);
        if (first_outside_idx_with_margin) {
          return *first_outside_idx_with_margin;
        }
        return *first_outside_idx;
      }();

      publishVirtualWall(optimized_traj_points.at(stop_idx).pose);
      for (size_t i = stop_idx; i < optimized_traj_points.size(); ++i) {
```

Comment (takayuki5168) [reply planning/obstacle_avoidance_planner/src/node.cpp:462]: Also, even if `enable_outside_drivable_area_stop_` is false, we wanna call `publishVirtualWall`.
So the following would be better.
```cpp
if (first_outside_idx) {
    const auto stop_idx = ...;
    publishVirtualWall(optimized_traj_points.at(stop_idx).pose);
    if (enable_outside_drivable_area_stop_) {
        for (size_t i = stop_idx; i < optimized_traj_points.size(); ++i) {
            ...
        }
    }
}
```

File: planning/obstacle_avoidance_planner/src/node.cpp (hunk 3 above)
Comment (takayuki5168) [reply planning/obstacle_avoidance_planner/src/node.cpp]: Thank you so much. I added some comments.

File: planning/obstacle_avoidance_planner/src/node.cpp (hunk 4 above)
Comment (ahmeddesokyebrahim) [reply planning/obstacle_avoidance_planner/src/node.cpp]: Thank you so much @takayuki5168 for the valuable comments :+1: .
I have updated the code, please have a look.

File: planning/obstacle_avoidance_planner/src/node.cpp [hunk 5]
```diff
@@ -434,16 +441,30 @@ void ObstacleAvoidancePlanner::insertZeroVelocityOutsideDrivableArea(
...
+  if (first_outside_idx) {
+    const auto stop_idx = [&]() {
+      const auto dist = tier4_autoware_utils::calcDistance2d(
+        optimized_traj_points.at(0), optimized_traj_points.at(*first_outside_idx));
+      const auto dist_with_margin = dist - vehicle_stop_margin_outside_drivable_area_;
+      const auto first_outside_idx_with_margin =
+        motion_utils::insertStopPoint(dist_with_margin, optimized_traj_points);
+      if (first_outside_idx_with_margin) {
+        return *first_outside_idx_with_margin;
+      } else {
```
Comment (takayuki5168) [inline planning/obstacle_avoidance_planner/src/node.cpp:459]: (Just comment. It's okay not to apply the change.)
This `else` can be removed. This is called early return.

**Review by takayuki5168 — APPROVED** [review]
Thank you for the change. I approve.

File: planning/obstacle_avoidance_planner/src/node.cpp (hunk 5 above)
Comment (ahmeddesokyebrahim) [reply planning/obstacle_avoidance_planner/src/node.cpp]: I have removed it. Thanks again for the comments!!

**Pull request merged**